src/plugins/discover/public/application/main/discover_main_route.tsx

```diff
@@ -260,27 +264,53 @@ export function DiscoverMainRoute({ customizationCallbacks, mode = 'standalone'
       },
       dataViewEditor,
       noDataPage: services.noDataPage,
-    };
+    }),
+    [core, data.dataViews, dataViewEditor, hasESData, hasUserDataView, services.noDataPage]
+  );
 
-    return (
-      <AnalyticsNoDataPageKibanaProvider {...analyticsServices}>
-        <AnalyticsNoDataPage onDataViewCreated={onDataViewCreated} />
-      </AnalyticsNoDataPageKibanaProvider>
-    );
-  }
+  const loadingIndicator = useMemo(
+    () => <LoadingIndicator type={hasCustomBranding ? 'spinner' : 'elastic'} />,
+    [hasCustomBranding]
+  );
+
+  const mainContent = useMemo(() => {
+    if (showNoDataPage) {
+      return (
+        <AnalyticsNoDataPageKibanaProvider {...noDataDependencies}>
+          <AnalyticsNoDataPage onDataViewCreated={onDataViewCreated} />
+        </AnalyticsNoDataPageKibanaProvider>
+      );
+    }
+
+    if (loading) {
+      return loadingIndicator;
```

Comment:
This change is expected. Originally it worked the way you're suggesting, but it was changed to align with how Log Explorer loads (with top nav visible), to allow switching tabs immediately instead of having to wait for Discover to load first, and to prevent hiding/showing the tabs whenever the user switches to Discover.

Before:

https://github.com/elastic/kibana/assets/25592674/30bf26d0-f9d8-40c9-bf7f-b228a93d47f6

Current (still a very minor flicker the first time the user clicks the Discover tab due to async components, but imperceptible after that):

https://github.com/elastic/kibana/assets/25592674/7c6c8c35-3225-4d41-9323-b633436cf4a5

I guess we'll have to pick between which UX tradeoff we prefer. Otherwise I think we'd need to make it so the tabs are hidden on the first load, but not on subsequent loads, which may get messy/difficult.

Reply:
Oh I see what you mean now! Yes we can and should hide those while loading, thanks. Updated here: https://github.com/elastic/kibana/pull/171467/commits/e7d57c2db89cc431759afffa2c07a1aabadc0745.